s3/src/main/java/org/togglz/s3/S3StateRepository.java

```diff
@@ -0,0 +1,118 @@
+package org.togglz.s3;
+
+import org.togglz.core.Feature;
+import org.togglz.core.repository.FeatureState;
+import org.togglz.core.repository.StateRepository;
+import org.togglz.core.util.FeatureStateStorageWrapper;
+
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.model.AmazonS3Exception;
+import com.amazonaws.services.s3.model.S3Object;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * A state repository that uses Amazon S3.
+ * <p>
+ * The repository is configured using the {@link Builder}
+ * <p>
+ * You must already have a bucket provisioned before you create this repository.
+ * 
+ * @author Mark Richardson
+ * @date 12/27/16
+ */
+public class S3StateRepository implements StateRepository {
+
+    // http://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html
+    private static final String ERR_NO_SUCH_KEY = "NoSuchKey";
+
+    private ObjectMapper objectMapper = new ObjectMapper();
+
+    protected final AmazonS3Client client;
+    protected final String bucketName;
+    protected final String keyPrefix;
+
+    private S3StateRepository(Builder builder) {

```

Comment:
Builder use should be optional. It is more typical approach for example with dependency injection.
This constructor should accept all fields as arguemnts, like in builder::build:
`return new S3StateRepository(client, bucketName, keyPrefix);`